File: sdks/kotlin/examples/SidecarChat.kt
import kotlinx.coroutines.flow.take
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking
import to.llmux.kotlin.LlmuxSidecar

/**
 * llmux SIDECAR from Kotlin — the gateway as a child process on 127.0.0.1.
 *
 * This is the recommended default for the JVM: no native library, no
 * --enable-native-access, no signal handlers, and it works on every platform
 * llmux builds for, Windows included.
 *
 * Runs on any Java 11+. Run it with sdks/kotlin/run-examples.sh.
 */
fun main() = runBlocking {
    // use {} stops the child process on every path out, rather than waiting for
    // the JVM shutdown hook.
    LlmuxSidecar().use { llmux ->
        println("sidecar: ${llmux.baseUrl}")
        println("openai base url: ${llmux.openAiBaseUrl}")

        println("models: ${llmux.models().oneLine(160)}")

        val request = """{"model":"demo","messages":[{"role":"user","content":"hello"}]}"""
        println("chat: ${llmux.chat(request).oneLine(200)}")

        // The same Flow shape the direct path offers, over SSE instead of a
        // callback. Swapping one for the other is a one-line change at the
        // construction site — that is the point of sharing the wire contract.
        val streamRequest =
            """{"model":"demo","messages":[{"role":"user","content":"hello"}],"stream":true}"""
        val chunks = llmux.chatChunks(streamRequest).toList()
        val text = chunks.joinToString("") { it.content() }
        println("stream: ${chunks.size} chunks, text=\"$text\"")

        // Cancelling unwinds through the `use` that owns the response body, so
        // the connection is closed. How many tokens the server had already
        // produced by then is not observable from here — unlike the direct
        // path, where the callback count is exact. That asymmetry is real and
        // worth knowing before you rely on early cancellation to save spend.
        val firstTwo = llmux.chatChunks(streamRequest).take(2).toList()
        println("early stop: collected ${firstTwo.size} chunk(s), response body closed")

        // The error path is an HTTP status, not an exception.
        val bad = llmux.chat("""{"model":"nope","messages":[]}""")
        println("error path: ${bad.oneLine(160)}")
    }
    println("sidecar stopped")
    println("done")
}

private fun String.content(): String {
    val needle = "\"content\":\""
    var i = indexOf(needle)
    if (i < 0) return ""
    i += needle.length
    val out = StringBuilder()
    while (i < length && this[i] != '"') {
        var c = this[i]
        if (c == '\\' && i + 1 < length) {
            i++
            c = if (this[i] == 'n') '\n' else this[i]
        }
        out.append(c)
        i++
    }
    return out.toString()
}

private fun String.oneLine(max: Int): String {
    val flat = replace(Regex("\\s+"), " ")
    return if (flat.length <= max) flat else flat.take(max) + "…"
}

File: sdks/dotnet/examples/SidecarChat.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Llmux;

namespace Llmux.Examples
{
    /// <summary>
    /// llmux SIDECAR — the gateway as a child process on 127.0.0.1, over HTTP.
    /// The recommended default for .NET.
    ///
    /// No native library, no unsafe code, no platform matrix — it works on
    /// Windows, where the direct path has no library at all.
    ///
    /// Run it with sdks/dotnet/run-examples.sh.
    /// </summary>
    internal static class SidecarChat
    {
        internal static async Task<int> RunAsync()
        {
            // try/finally rather than `using`: the sidecar is a process-wide
            // singleton owned by the static class, and it must be stopped on the
            // error path too rather than left for the exit hook.
            string baseUrl = Sidecar.BaseUrl();
            using var http = new HttpClient { Timeout = TimeSpan.FromSeconds(30) };
            try
            {
                Console.WriteLine($"sidecar: {baseUrl}");
                Console.WriteLine($"openai base url: {Sidecar.OpenAIBaseUrl()}");

                // 1. The model list.
                string models = await http.GetStringAsync(baseUrl + "/v1/models");
                Console.WriteLine($"models: {DirectChat.OneLine(models, 160)}");

                // 2. One completion.
                const string request =
                    """{"model":"demo","messages":[{"role":"user","content":"hello"}]}""";
                string chat = await PostAsync(http, baseUrl, request);
                Console.WriteLine($"chat: {DirectChat.OneLine(chat, 200)}");

                // 3. Streaming, as the same IAsyncEnumerable shape the direct
                //    path offers — over SSE instead of a callback.
                const string streamRequest =
                    """{"model":"demo","messages":[{"role":"user","content":"hello"}],"stream":true}""";
                var text = new StringBuilder();
                int chunks = 0;
                await foreach (string chunk in StreamAsync(http, baseUrl, streamRequest))
                {
                    chunks++;
                    text.Append(DirectChat.Content(chunk));
                }
                Console.WriteLine($"stream: {chunks} chunks, text=\"{text}\"");

                // 4. Breaking out disposes the response stream, which drops the
                //    connection. How much the server had already generated is
                //    NOT observable from here — unlike the direct path, where
                //    the callback count is exact.
                int consumed = 0;
                await foreach (string chunk in StreamAsync(http, baseUrl, streamRequest))
                {
                    consumed++;
                    if (consumed == 2)
                    {
                        break;
                    }
                }
                Console.WriteLine($"early stop: consumed {consumed} chunk(s), response disposed");

                // 5. The error path is a status code, not an exception.
                string bad = await PostAsync(http, baseUrl, """{"model":"nope","messages":[]}""");
                Console.WriteLine($"error path: {DirectChat.OneLine(bad, 160)}");
            }
            finally
            {
                Sidecar.Stop();
                Console.WriteLine("sidecar stopped");
            }

            Console.WriteLine("done");
            return 0;
        }

        private static async Task<string> PostAsync(HttpClient http, string baseUrl, string body)
        {
            using var content = new StringContent(body, Encoding.UTF8, "application/json");
            using HttpResponseMessage response =
                await http.PostAsync(baseUrl + "/v1/chat/completions", content);
            return await response.Content.ReadAsStringAsync();
        }

        /// <summary>SSE `data:` frames as chunk JSON, with `[DONE]` stripped.</summary>
        private static async IAsyncEnumerable<string> StreamAsync(
            HttpClient http, string baseUrl, string requestJson,
            [EnumeratorCancellation] CancellationToken cancellationToken = default)
        {
            using var request = new HttpRequestMessage(
                HttpMethod.Post, baseUrl + "/v1/chat/completions")
            {
                Content = new StringContent(requestJson, Encoding.UTF8, "application/json"),
            };
            request.Headers.Accept.ParseAdd("text/event-stream");

            // ResponseHeadersRead: do not buffer the whole body before yielding
            // the first chunk, or "streaming" is a word rather than a behaviour.
            using HttpResponseMessage response = await http.SendAsync(
                request, HttpCompletionOption.ResponseHeadersRead, cancellationToken);
            using Stream body = await response.Content.ReadAsStreamAsync(cancellationToken);
            using var reader = new StreamReader(body, Encoding.UTF8);

            while (!reader.EndOfStream)
            {
                string? line = await reader.ReadLineAsync(cancellationToken);
                if (line == null)
                {
                    break;
                }
                if (!line.StartsWith("data: ", StringComparison.Ordinal))
                {
                    continue;
                }
                string payload = line.Substring("data: ".Length);
                if (payload == "[DONE]")
                {
                    break;
                }
                yield return payload;
            }
        }
    }
}
